doc/tutorial/tutorial.md

```diff
@@ -1859,23 +1857,133 @@ Output `multiset`s are more memory-efficient than `relation`s: internally the
 implementation uses only multisets, and converts them to sets when needed using
 an expensive Differential Dataflow `distinct` operator.
 
-Streams are yet another kind of relation in DDlog that are similar to multisets
-with one additional optimization.  Normally, DDlog stores a copy of the entire
-contents of an `input relation` or `input multiset`.  This copy is used to
-implement the `clear` command, which removes everything from the relation.  In
-addition, it is necessary to enforce the set semantics on `relation`s.  A stream
-is a multiset, whose contents is not cached by DDlog, thus reducing the memory
-footprint of the program.  As a result, it is illegal to use the `clear` command
-on a stream.
+## Streams
+
+Incremental computation is often used in stream analytics in order to update
+results of a query as new data arrives in real time.  Such applications
+typically operate on a mix of persistent data that, once inserted in a relation,
+remains there until explicitly deleted, and **ephemeral data** that gets
+processed and instantly discarded.
+
+As a case in point, the `ZipCodes` relation below, that maps a zip code to the
+city where this zip is located, only changes occasionally.  In contrast, the
+`Parcels` relation, that contains information about parcels received by the
+postal service, may ingest thousands of records per second.  We join the two
+relations and output a copy of records from `Parcels` enriched with the name of
+the city where each parcel must be sent.
+
+```
+input relation ZipCodes(zip: u32, city: string)
+input relation Parcel(zip: u32, weight: usize)
+
+// Add the name of the destination city to each parcel.
+output relation ParcelCity(zip: u32, city: string, weight: usize)
+ParcelCity(zip, city, weight) :-
+    Parcel(zip, weight),
+    ZipCodes(zip, city).
+```
+
+At runtime, the client inserts a bunch of records in the `Parcel` relation,
+reads the corresponding outputs from `ParcelCity` and clears the `Parcel`
+relation since its contents is no longer needed and to avoid running out of
+memory.  This is cumbersome and expensive.
+
+A better solution is to use the DDlog **stream** abstraction, intended to
+simplify working with ephemeral data, i.e., data that gets discarded after being
+used.  A stream is a relation that only contains records added during the
+current transaction.  When the transaction completes, all previously inserted
+records as well as any records derived from them disappear without a trace and
+the memory footprint of all streams in the program drops to zero.
+
+We re-declare the `Parcel` relation as a stream (by using the `stream` keyword
+instead of `relation`).  Any rule that uses a stream yields a stream; hence
```

Comment:
This seems incredibly limiting, since streams also can't be used in recursive rules and such this makes streams only applicable to a very small subset of use cases

Reply:
The whole point of streams is that they have zero memory footprint.  Recursive computations in DD fundamentally require indexing stuff, so this is not an easy limitation to lift.  Having said that, you can have as much recursion as you like in your program, and then join the result of the recursive computation with a stream.  We also plan to add more things you can do with streams in the near future.  For example, we might support rolling up a stream into a "normal" relation that you can then use with recursion, etc.

Reply:
PS. Limited as they are, streams speed things up a lot when they are applicable. We have some internal use cases that benefit from this.